### ISSUE TITLE
Merge chai-as-promised into core, as Promise now are widely used.

### ISSUE THREAD

**Issue by lygstate** [chaijs/chai#882]

**Comment by lucasfcosta** [comment]
Hi @lygstate, thanks for your issue, this does make total sense.

Maybe we could do this as we did with `chai-change`, but first we'd need to talk to the whole team and specially with @domenic, which is `chai-as-promised` author (and did an awesome job with it BTW).

Personally I feel like the tendency of any plugin or library that gets widely used is to become part of the core and this does not happen only with Chai. A clear example for that is what happened with jQuery's selectors and `document.querySelector`.

Anyway, let's wait to hear from everyone before taking a decision.

**Comment by astorije** [comment]
I agree that this makes a lot of sense indeed, but there is something to take into account here.

Since `chai-as-promised` works by [overriding available assertions](https://github.com/domenic/chai-as-promised/blob/b2cfbdc71360dad1faaa29f64bcc8ba54819084e/lib/chai-as-promised.js#L241-L255), plugins will be affected.

For example, [`chai-immutable`](https://github.com/astorije/chai-immutable) works by overriding assertions so that when evaluating an [Immutable](https://github.com/facebook/immutable-js) structure, the plugin gets used, otherwise it defaults to the existing assertions.
To be compatible with `chai-as-promised`, plugins overriding methods need to be loaded **before** `chai-as-promised`, as explained [here](https://github.com/astorije/chai-immutable#using-chai-immutable-with-other-plugins) (see https://github.com/astorije/chai-immutable/issues/32).

While I'd personally be happy to see `chai-as-promised` in the main project, and apply necessary fixes to `chai-immutable` accordingly, this would break existing plugins and require updating [documentations of the `override*(...)` plugin utilities](http://chaijs.com/api/plugins/).

**Comment by lucasfcosta** [comment]
Thanks for the warning, @astorije.

Wouldn't it be possible to make sure we "reloaded" `chai-as-promised` after each time someone called the `use` method?

Maybe we could keep track of the methods being added whenever someone calls any of the `add*(...)` or `overwrite(...)` methods and then override those again with our internal implementation of `chai-as-promised` in order to guarantee they would also be overwritten.

I didn't do much exploring on `chai-as-promised`'s code, so let me know if I said anything wrong/impossible.

**Comment by astorije** [comment]
I'm not sure about wrong/impossible as I am not an expert with `chai-as-promised` code, we'd need @domenic to chime in :)

However, I'm seeing a potential issue if some plugins expect to run after `chai-as-promised` or any plugin that affects existing assertions. I'm thinking about plugins like https://github.com/prodatakey/dirty-chai (odd that this plugin didn't appear in the plugins page, though I'm thinking this one will become unnecessary after v4).

I actually don't know `chai-as-promised` could be integrated as is in the repo. If I remember correctly, it works by adding a `eventually` flag then monkey-patching all assertions to support it. Merging it with the main project would instead implement support for `eventually` natively in assertions, and therefore require all plugins to do so as well. Either that or transform `chai-as-promised` into `plugin-as-promised` that monkey patches plugins only...
I don't know, I might be very wrong, I'm tired :) I'd love to hear @domenic's thoughts either way.

**Comment by demurgos** [comment]
What is the current state of this issue? I believe that there was not much change?
After traking the existing issues, I see the following highlights:

- [2015-04-07]: @keithamus asks @domenic  for a migration to core
- [2016-01-30](https://github.com/chaijs/chai/issues/605): Chai wants to support Promises since they are JS primitives
- [2016-10-22](https://github.com/chaijs/chai/issues/415#issuecomment-255549627): Add Promise support to handle ES7 async functions
- 2017-05-26: Chai@4.0.0

Is it possible that first-class support for promises will land in the 4.x version?

**Comment by keithamus** [comment]
Good summary of events @demurgos.

I think perhaps async functions are the way future. I personally think Promises will eventually become an implementation detail of JavaScript's async functions; and testing will be no different. Async functions using Mocha/Chai are totally doable today:

```js
const myPromisedValue = Promise.resolve(4)
const otherPromisedValue = Promise.resolve(['a', 'b'])
it('works with async', async () => {
  expect(await myPromisedValue).to.equal(4)
  expect(await otherPromisedValue).to.be.an('array').that.deep.equals(['a', 'b'])
});
```

I super appreciate chai-as-promised, it is certainly one of the "defacto" plugins for chai, and - were it not for async functions - I'd absolutely be trying harder to get it in core. However I think with a bit more time, we might find that async functions render chai-as-promised a little redundant.

**Comment by demurgos** [comment]
Thank you, for your answer.
You are right that it's pretty convenient to chai with use `await` to check for equality, but it does not cover cases when the function is invoked by Chai. It happens for assertions such as "throws" or "changes".
It would be nice if Chai provided an async variant for example, maybe something like:

```js
async function foo() {throw new Error("Foo");}
it("`foo` throws an async error (rejected Promise)", async () => {
  await assert.throwsAsync(foo);
});
```

**Comment by keithamus** [comment]
Good point. Not every part of the async story works well with chai. I think we should look to provide mechanisms for that, like you say. We need to think carefully though; make sure that we do it the right way.

**Comment by demurgos** [comment]
Nice to hear that it's still planned and you're thinking about it. 🙂
 
I agree that it shouldn't be rushed. Even if I would be happy with an async variant of throws, the example I provided can cause some issues for newcomers if they forgot the "await" keyword since the promise resolution will happen outside of the test context. It's not a problem specific to tests, just to await and Promises in general, I don't know if there's a way to solve it.

**Comment by keithamus** [comment]
> newcomers if they forgot the "await" keyword since the promise resolution will happen outside of the test context

Precisely the problem we have - and actually `chai-as-promised` suffers from a similar problem (in that you need to return the assertion to mocha for it to work effectively - which is not idiomatic chai). One of the big issues for newcomers to chai is that chai silently passes and only noisily fails - which means if you, for example, write `expect(false).to.be.ture` it'll silently pass. We've worked towards many mitigation strategies for these problems in Chai 4, to ensure newcomers aren't left with invalid passing tests, and I'm sure the chai team - especially @meeber (who spent a great and noble effort on this) does not want to repeat that mistake with Promises.

Discussion on how we solve this is completely open by the way; if you have suggestions please feel free to share. It is important to note though that we may not take them on board due to various semantics about the design/usability of chai. For example `throwsAsync` sounds perfectly reasonable, excluding the above points - so we should look for a "more robust" solution than this, I think.

**Comment by demurgos** [comment]
After thinking about it, this problem is inherent to async programming, but if you want to solve this in the context of chai assertions, you could force the promises to be handled.

Bluebird and when allow you to detect "possibly unhandled promise rejections". They keep a reference count of Promise observers (incremented by each `.then` call). If the promise is rejected and there is no observer, it throws this exception. This prevents late observers: when you attach an observer _after_ the promise was resolved but it's usually a good trade-off.

For Chai, you could use a similar approach. Use a sub-class of Promise with hooks on the `.then` and `.catch` methods as the return value of your async assertions. If this promise is fulfilled (resolved or rejected) when it has no observer (or one tick after), consider it as a "dangling Promise" that wasn't awaited. Since your goal is to prevent this sort of Promise and that it will be used in the context of tests anyway, print an error in the console, throw the error or even kill the process: get the attention of the developer and tell them that the result reported by Mocha may be false because some async assertions weren't awaited.

**Comment by keithamus** [comment]
One solution could be to make the assertion chain a `thenable` so that it _always_ has a `then` method which would do what you expect - effectively mimicking Promises. The problem though is much more of a usability one that anything else. Not returning a Promise causes all kinds of weirdness, for both passing and failing tests; and it can be a difficult fix to spot. We, as a library, do not have the hooks into the users environment to tell them something like 'you missed a return statement on line 20' and so the usability barrier is left with the developer. If we can somehow solve that then we can progress with native promise support.

**Comment by grnch** [comment]
Hi @keithamus.

I read the entire discussion above, but I was left somewhat unclear on what you consider as the problem that needs to be solved before integrating native promise support further into Chai. I would really appreciate if you could clarify a bit more, for the rest of us mortals. :)

Maybe a concrete example would help to highlight the issue.

Imagine we are working in a native ES7 environment, and we have these two async functions:

```javascript
async function good() {
    return "good";
}

async function bad() {
    throw new Error("bad");
}
```

Now consider these two Chai assertions one might make about these functions:

```javascript
expect(await good()).to.equal("good");

await expect(bad()).to.throwAsync("bad");
```

(the `throwAsync()` is a hypothetical assertion that takes a Promise and verifies that it ultimately gets rejected with a given error)

1) Is the problem here the inconsistency with the placement of the `await` keyword? In the first test the `await` keyword needs to be inside the `expect()`, yet in the second test the `await` needs to be outside `expect()`.

2) Is it the fact that if you forget the `await` keyword in the first test it will loudly fail, whereas if you forget the `await` on the second test it will always silently pass, regardless if the assertion was successful or not?

3) Or is it something else entirely?

**Comment by keithamus** [comment]
@grnch 

Just to set the record straight - `expect(async good()).to.equal('good')` would work right now, in chai 4, chai 3, probably even right back to chai 0.1. Specifically; the pattern we are lacking is `.to.throw()` on async functions.

Theoretically, nothing is wrong with the above code examples - `.throwAsync` as per your example _could_ work fine! The big issue we have is usability. It's a softer thing than technical feasibility. I'll explain...

Chai is quite "opaque" then it comes to what it returns; the intuitive thing for an `expect()` call is to do nothing with the return value (because a bad assertion throws rather than having a meaningful return value). We see lots of users trip up with this with chai-as-promised; have a [read through some of the closed issues on chai-as-promised](https://github.com/domenic/chai-as-promised/issues?utf8=%E2%9C%93&q=is%3Aclosed%20return%20mocha%20), most of them are usability issues.

So; yes, today, we _could_ have the following, with a hypothetical new `throwAsync` method:

```js
describe('myasyncfn', () => {
  //async
  it('throws with wrong arguments', async () => {
    await expect(bad()).to.throwAsync('bad')
  })
  //return promise
  it('throws with wrong arguments', () => {
    return expect(bad()).to.throwAsync('bad')
  })
  //implicitly returning concise arrow
  it('throws with wrong arguments', () =>
    expect(bad()).to.throwAsync('bad')
  )
})
```

But then all of these would not work:

```js
describe('myasyncfn', () => {
  //await inside expect
  it('DOES NOT WORK: throws with wrong arguments', async () => {
    expect(await  bad).to.throwAsync('bad')
  })
  //await wrapped expect
  it('DOES NOT WORK: throws with wrong arguments', async () => {
    await (expect(bad)).to.throwAsync('bad')
  })
  //dont return
  it('DOES NOT WORK: throws with wrong arguments', () => {
    expect(bad).to.throwAsync('bad')
  })
  //using throw on async fn
  it('DOES NOT WORK: throws with wrong arguments', async () => {
    await expect(bad).to.throw('bad')
  })
  //using throw on async fn with return
  it('DOES NOT WORK: throws with wrong arguments', () => {
    return expect(bad).to.throw('bad')
  })
  //using throw on async fn with return with implicit returning concise arrow
  it('DOES NOT WORK: throws with wrong arguments', () =>
    expect(bad).to.throw('bad')
  )
})
```
As you can see all of these functions are _very_ subtly different but break in big ways - this is _not_ what we want with Chai, things need to loudly break in obvious ways. No one (not even the chai maintainers!) wants to spend time debugging silly mistakes like typos or wrong method calls.

The worst part about this is that many of the functions above would _silently pass_ as neither Mocha nor Chai know anything about the callee code, and cannot determine when you've triggered a Promise that you're meant to return.

Now, there are ways around some of these - I won't get into them now because it's tangential to the point; right now - there is no simple, intuitive way to handle async errors. Fundamentally, it'll probably require chai's `expect()` return value to be a thenable, and that is a new pattern for chai, and new patterns means a new category of problems.

I'll digress a little; and talk a bit more personally here - hopefully this doesn't come off as whiney. Ultimately this all comes down to time. In a world of infinite time I'd implement this feature right away, rigorously document it, and also spend the hours and hours of time responding to users who get tripped up on the issues (and also deal with occasional "you're an idiot I wasted hours of my life on your bad code" comments - sadly, they happen). In reality any time developing this (kind of small) feature would actively take time away from us developing much bigger or higher priority features that would impact more users (and alleviate more of our time from support requests). No one on the chai team works on this full time, and we don't have any budget to hire anyone to do so (despite it being one open source project that is used much more in the corporate world than the hobbyist world). I guess what I'm saying is - it's about incentives; I could spend a day writing `throwAsync`, and in return receive a bunch more issues in our issue tracker, or I could spend that day improving our inspection engine, and in return all of the issues around inspect not working properly die down.

**TL;DR**: writing a function like `throwsAsync` would introduce subtle problems into users codebases, and we (the maintainers) would spend more time responding to those users than developing features. We'd rather let the dust settle on async functions a bit more, get some higher priority features out of the way, and some day later we can revisit this and try to come up with a more usable solution.

**Comment by grnch** [comment]
@keithamus 

Thank you for your detailed and thoughtful response. Much appreciated.

That list of potential pitfalls you posted was especially illuminating!

There may be some technical tricks to soften the impact of these pitfalls, but at the end of the day, modern JavaScript is a very flexible language with a lot of syntax variations and semantic subtleties. It's pretty easy for JS developers to shoot themselves in the foot even without Chai, so it may be too ambitious to expect Chai to address all the shortcomings of the language. :)

However, I do appreciate your candid reflection on the realities of an all-volunteer open-source project. Of course you have to prioritize things and spend your limited time on things that offer the most value to the project - no argument there. It's also perfectly understandable that you might want to wait until this whole ES7 async/await thing settles down, and best practices start to emerge around it.

Besides, as you pointed out, Chai can already handle Promise values just fine an ES7 environment. All you really have to do is put an `await` keyword before any Promise, and then you can use the rest of Chai machinery just fine on the resulting value.

The only real gap is that the existing `throw()` assertion cannot handle async functions that throw async exceptions (i.e. Promise rejections).

Perhaps the way forward here is to just create a "chai-throw-async" plugin that introduces this new `throwAsync()` assertion, to see if this approach can address the gap in a satisfactory way.

It would be much less ambitious and much smaller in scope than something like "chai-as-promised", which should hopefully make it much easier to merge into Chai core if it ever gains enough traction with developers.

This should allow people to test out this approach and shake out the problems without cluttering the main Chai issue tracker. Any bugs or problems would hopefully go to the plugin issue tracker instead.

I'll think about this some more, and maybe try my hand at creating such a plugin (never created a Chai plugin before).

**Comment by keithamus** [comment]
@grnch creating a plugin sounds like a fantastic idea! One for the big selling points for plugins is they can experiment much more outside of core, and we are always happy to fold plugins into core - where it makes sense. I'd very much encourage you do go forth and write plugins! If you do, please give us some feedback on how plugins could be better handled, or better documented!

**Comment by alfonsorios96** [comment]
It's very useful for the developers this change. I want to chai-as-a-promised into core because it's easier implement the new features and a improvement for create new functionality. The promises are very important for javascript.

**Comment by machineghost** [comment]
>I guess what I'm saying is - it's about incentives; I could spend a day writing throwAsync, and in return receive a bunch more issues in our issue tracker, or I could spend that day improving our inspection engine, and in return all of the issues around inspect not working properly die down.

This is just a single data-point from an outsider, and please don't take it as an attack (anymore than your post was a whine), but I just want to point out that fixing your engine can amount to re-arranging the deck chairs on the Titanic if you lose enough mindshare in the process of doing it.

I'm a developer who's used Chai for years, but when I started re-using it for a new project I realized:

* I need a plugin for strings
* I need a plug-in for promises
* I need a custom-written `throwAsync` plugin that doesn't even exist yet to test `await`/`async`

 ... and all of that is *just* for stuff that's baked in to the Javascript language itself (I'm not even talking about stuff like `chai-enyzme` ).

Alternatively I can use Hapi: one `npm i` command and I'm done.  When you ignore tickets like this one in favor of perfecting your engine I think you run the real risk of devs following my train of thought and abandoning the library entirely in favor of one that does include the batteries.

**Comment by keithamus** [comment]
I'm going to close this. We have thought about this a lot and there's been great discussion here but I think ultimately we might look at adding a `.rejected` style assertion but async/await will subsume any need for chai as promised.
